manager/controllers/app/storage.go

```diff
@@ -6,168 +6,95 @@ package app
-func FindAvailableBucket(c client.Client, log logr.Logger, owner types.NamespacedName, id string, prefixBase string, canShare bool, geo string) *app.M4DBucket {
-	ctx := context.Background()
+// ProvisionInterface is an interface for managing Dataset resources
+type ProvisionInterface interface {
+	CreateDataset(dataset *comv1alpha1.Dataset) error
+	//DeleteDataset(namespaced types.NamespacedName) error
+	//GetStatus(namespaced types.NamespacedName) (*comv1alpha1.DatasetStatus, error)
+}
 
-	var buckets app.M4DBucketList
-	log.Info("Searching for an available bucket")
-	if err := c.List(ctx, &buckets); err != nil {
-		log.Info(err.Error())
-		return nil
-	}
-	for _, bucket := range buckets.Items {
-		utils.PrintStructure(bucket, log, "Bucket ")
-		if IsBucketAvailable(&bucket, owner, id, canShare) {
-			AddOwner(&bucket, owner)
-			GetOrCreatePrefix(&bucket, id, prefixBase)
-			if err := c.Status().Update(ctx, &bucket); err != nil {
-				// the object has been updated by someone else - continue to the next one
-				if statusErr.IsConflict(err) {
-					log.Info("Conflict during an update of M4DBucket " + bucket.Name)
-				}
-				log.Info("Could not update M4DBucket " + bucket.Name)
-				continue
-			}
-			return &bucket
-		}
-	}
-	return nil
+type ProvisionImpl struct {
+	Client client.Client
 }
 
-// FreeStorageAssets removes the app identifier from the list of owners
-func (r *M4DApplicationReconciler) FreeStorageAssets(owner types.NamespacedName) error {
-	ctx := context.Background()
-	var buckets app.M4DBucketList
-	if err := r.List(ctx, &buckets); err != nil {
-		return err
-	}
-	for _, bucket := range buckets.Items {
-		if !HasOwner(&bucket, owner) {
-			continue
-		}
-		RemoveOwner(&bucket, owner)
-		if !HasOwners(&bucket) {
-			bucket.Status.AssetPrefixPerDataset = map[string]string{}
-		}
-		if err := r.Client.Status().Update(ctx, &bucket); err != nil {
-			r.Log.V(0).Info("Error during an update of M4DBucket: " + err.Error())
-			// the object has been updated by someone else - continue to the next one
-			if statusErr.IsConflict(err) {
-				r.Log.V(0).Info("Conflict during an update of M4DBucket " + bucket.Name)
-				return err
-			}
-		}
+func NewProvisionImpl(c client.Client) *ProvisionImpl {
+	return &ProvisionImpl{
+		Client: c,
 	}
-	return nil
 }
 
-// Utility functions
-
-// CreateOwnerID creates a string based on namespace and name values
-func CreateOwnerID(owner types.NamespacedName) string {
-	return owner.Namespace + "/" + owner.Name
+func (r *ProvisionImpl) CreateDataset(dataset *comv1alpha1.Dataset) error {
+	return r.Client.Create(context.Background(), dataset)
 }
 
-// HasOwner checks whether the given owner owns the resource
-func HasOwner(resource *app.M4DBucket, owner types.NamespacedName) bool {
-	ownerID := CreateOwnerID(owner)
-	for _, val := range resource.Status.Owners {
-		if val == ownerID {
-			return true
-		}
-	}
-	return false
+type ProvisionFake struct {
```

Comment:
Why is this called fake?

Reply:
It is used for testing, not a real provisioning. I can change the name...